server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java

```diff
@@ -446,55 +455,85 @@ private Query getVectorQuery(QueryShardContext context, ShapeBuilder queryShapeB
         }
 
         // wrap geoQuery as a ConstantScoreQuery
-        return getVectorQueryFromShape(context, queryShapeBuilder.buildLucene());
-    }
-
-    private Query getVectorQueryFromShape(QueryShardContext context, Object queryShape) {
-        Query geoQuery;
-        if (queryShape instanceof Line[]) {
-            geoQuery = LatLonShape.newLineQuery(fieldName(), relation.getLuceneRelation(), (Line[]) queryShape);
-        } else if (queryShape instanceof Polygon[]) {
-            geoQuery = LatLonShape.newPolygonQuery(fieldName(), relation.getLuceneRelation(), (Polygon[]) queryShape);
-        } else if (queryShape instanceof Line) {
-            geoQuery = LatLonShape.newLineQuery(fieldName(), relation.getLuceneRelation(), (Line) queryShape);
-        } else if (queryShape instanceof Polygon) {
-            geoQuery = LatLonShape.newPolygonQuery(fieldName(), relation.getLuceneRelation(), (Polygon) queryShape);
-        } else if (queryShape instanceof Rectangle) {
-            Rectangle r = (Rectangle) queryShape;
-            geoQuery = LatLonShape.newBoxQuery(fieldName(), relation.getLuceneRelation(),
-                r.minLat, r.maxLat, r.minLon, r.maxLon);
-        } else if (queryShape instanceof double[][]) {
-            // note: we decompose point queries into a bounding box query with min values == max values
-            // to do this for multipoint we would have to create a BooleanQuery for each point
-            // this is *way* too costly. So we do not allow multipoint queries
-            throw new QueryShardException(context, "Field [" + fieldName + "] does not support " + GeoShapeType.MULTIPOINT + " queries");
-        } else if (queryShape instanceof double[] || queryShape instanceof GeoPoint) {
-            // for now just create a single bounding box query with min values == max values
-            double[] pt;
-            if (queryShape instanceof GeoPoint) {
-                pt = new double[] {((GeoPoint)queryShape).lon(), ((GeoPoint)queryShape).lat()};
-            } else {
-                pt = (double[])queryShape;
-                if (pt.length != 2) {
-                    throw new QueryShardException(context, "Expected double array of length 2. "
-                        + "But found length " + pt.length + " for field [" + fieldName + "]");
+        return getVectorQueryFromShape(context, queryShapeBuilder.buildGeometry());
+    }
+
+    private Query getVectorQueryFromShape(QueryShardContext context, Geometry queryShape) {
+        return queryShape.visit(new GeometryVisitor<Query>() {
```

Comment:
This one isn't so bad since the only thing this method does is instantiate a new GeometryVisitor.